server/entity/physics/trace/trace.go

```diff
@@ -0,0 +1,103 @@
+package trace
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+)
+
+// TraverseBlocks performs a ray trace between the start and end coordinates.
+// A function 'f' is passed which is called for each voxel, if f returns false, the function will return.
+// TraverseBlocks panics if the start and end positions are the same.
+func TraverseBlocks(start, end mgl64.Vec3, f func(pos cube.Pos) (con bool)) {
+	dir := end.Sub(start).Normalize()
+	if dir.LenSqr() <= 0.0 {
+		panic("start and end points are the same, giving a zero direction vector")
+	}
+
+	b := cube.PosFromVec3(start)
+
+	step := signVec3(dir)
+	stepX, stepY, stepZ := int(step[0]), int(step[1]), int(step[2])
+	max := boundaryVec3(start, dir)
+
+	delta := divideVec3(step, dir)
+
+	r := world.Distance(start, end)
+	for {
+		if !f(b) {
+			return
+		}
+
+		if max[0] < max[1] && max[0] < max[2] {
+			if max[0] > r {
+				return
+			}
+			b[0] += stepX
+			max[0] += delta[0]
+		} else if max[1] < max[2] {
+			if max[1] > r {
+				return
+			}
+			b[1] += stepY
+			max[1] += delta[1]
+		} else {
+			if max[2] > r {
+				return
+			}
+			b[2] += stepZ
+			max[2] += delta[2]
+		}
+	}
+}
+
+// divideVec3 ...
+func divideVec3(v1, v2 mgl64.Vec3) mgl64.Vec3 {
+	return mgl64.Vec3{divide(v1[0], v2[0]), divide(v1[1], v2[1]), divide(v1[2], v2[2])}
+}
+
+// divide ...
+func divide(f1, f2 float64) float64 {
```

Comment:
Would maybe call this `safeDivide` or something like that. (Same for the vec3 version above)